src/audio/detect_test.c

```diff
@@ -35,17 +35,138 @@
 #include <sof/stream.h>
 #include <sof/ipc.h>
 #include <sof/audio/component.h>
+#include <uapi/user/detect_test.h>
 
 /* tracing */
-#define trace_keyword(__e) trace_event(TRACE_CLASS_MUX, __e)
-#define trace_keyword_error(__e)   trace_error(TRACE_CLASS_MUX, __e)
-#define tracev_keyword(__e)        tracev_event(TRACE_CLASS_MUX, __e)
+#define trace_keyword(__e, ...) \
+	trace_event(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+#define trace_keyword_error(__e, ...) \
+	trace_error(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+#define tracev_keyword(__e, ...) \
+	tracev_event(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+
+#define DETECT_TEST_SPIKE_THRESHOLD 0x00FFFFFF
 
 struct comp_data {
+	uint32_t source_period_bytes;	/**< source number of period bytes */
+	uint32_t sink_period_bytes;	/**< sink number of period bytes */
+	enum sof_ipc_frame source_format;	/**< source frame format */
+	enum sof_ipc_frame sink_format;	/**< sink frame format */
 	uint32_t period_bytes;
 	uint32_t switch_state[2];
+
+	struct sof_detect_test_config *config;
+	void *load_memory;	/**< synthetic memory load */
+	int32_t *prev_samples;	/**< last samples from previous period */
+
+	void (*detect_func)(struct comp_dev *dev, struct comp_buffer *sink,
+			    struct comp_buffer *source, uint32_t frames);
 };
 
+static void detect_test_notify(struct comp_dev *dev)
+{
+	struct sof_ipc_comp_event event;
+
+	trace_keyword("detect_test_notify()");
+
+	event.event_type = SOF_CTRL_EVENT_KD;
+	event.num_elems = 0;
+
+	ipc_send_comp_notification(dev, &event);
+}
+
+static void default_detect_test(struct comp_dev *dev, struct comp_buffer *sink,
+				struct comp_buffer *source, uint32_t frames)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	int32_t *src = source->r_ptr;
+	int32_t *dest = sink->w_ptr;
+	uint32_t count = frames * dev->params.channels;
+	uint32_t sample;
+	int notified = 0;
+
+	/* synthetic load */
+	if (cd->config)
+		idelay(cd->config->load_mips * 1000000);
+
+	/* compare with samples from previous period */
+	for (sample = 0; sample < dev->params.channels; ++sample) {
+		if (abs(cd->prev_samples[sample] - src[sample]) >=
+		    DETECT_TEST_SPIKE_THRESHOLD) {
+			detect_test_notify(dev);
+			notified = 1;
+		}
+	}
+
+	/* copy the samples and perform detection within current perioud */
+	for (sample = 0; sample < count; ++sample) {
+		dest[sample] = src[sample];
+
+		if (!notified && sample > 0 &&
+		    abs(src[sample - 1] - src[sample]) >=
+		    DETECT_TEST_SPIKE_THRESHOLD) {
```

Comment:
should we compare with the same channel only? e.g. for 2 channels data, compare sample 0 with 2, 4, ... and sample 1 with 3, 5, ...?

Reply:
Oops, good catch